dashboard/modules/log/log_agent.py

```diff
@@ -252,6 +289,62 @@ async def ListLogs(self, request, context):
             log_files.append(p.name)
         return reporter_pb2.ListLogsReply(log_files=log_files)
 
+    async def _find_task_log_offsets(
+        self, task_id: str, attempt_number: int, lines: int, f: io.BufferedIOBase
+    ) -> Tuple[int, int]:
+        """Find the start and end offsets in the log file for a task attempt
+        Current task log is in the format of below:
+
+            :job_id:xxx
+            :task_name:xxx
+            :task_attempt_start:<task_id>-<attempt_number>
+            ...
+            actual user logs
+            ...
+            :task_attempt_end:<task_id>-<attempt_number>
+            ... (other tasks)
+
+
+        For async actor tasks, task logs from multiple tasks might however
+        be interleaved.
+        """
+        # Find the end of the task log, which is the start of the next task log if any
+        # with the LOG_PREFIX_TASK_ATTEMPT_END magic line.
+        task_attempt_end_magic_line = (
+            f"{LOG_PREFIX_TASK_ATTEMPT_END}{task_id}-{attempt_number}\n"
+        )
+        end_offset = find_offset_of_content_in_file(
+            f, task_attempt_end_magic_line.encode()
+        )
+        if end_offset == -1:
+            # No other tasks (might still be running), stream til the end.
+            end_offset = find_end_offset_file(f)
+
+        # Find start
+        task_attempt_start_magic_line = (
+            f"{LOG_PREFIX_TASK_ATTEMPT_START}{task_id}-{attempt_number}\n"
+        )
+        task_attempt_magic_line_offset = find_offset_of_content_in_file(
```

Comment:
how expensive it is to search offset? Should we combine two of them into a single scan?

Reply:
Rearrange to locate start first, and then end.